File: internal/web/static/js/testdata/library_allmodes_test.js
'use strict';
// Runs the LITERAL mvrimport.js the browser loads, and asserts that every
// GDTF entering Benny512 puts its FULL mode list into the Fixture Library.
//
// Why this test loads the real file rather than reimplementing the shape:
// the defect it guards is two hand-rolled copies of one document drifting
// apart. library.js used to build the library record inline; patch.js now
// imports GDTFs too. A test that built its own expected record would agree
// with whichever copy it was written from and prove nothing about the other.
//
// The rule being pinned: a patch entry stores ONE mode (patch.Entry: Mode,
// Footprint, ChannelFunctions). Import is therefore the only moment a
// fixture's other modes exist — harvesting the show afterwards can only ever
// return the one mode each entry actually uses. A GDTF import that keeps
// just the applied mode has silently discarded the rest.

const fs = require('fs');
const path = require('path');
const vm = require('vm');

const jsDir = path.resolve(__dirname, '..');

let failures = 0;
function check(cond, msg) {
  if (!cond) { failures++; console.error('FAIL: ' + msg); }
}

// --- load mvrimport.js with just enough of its world -----------------------
// It reaches for MvrZip, GdtfParse and Api at call time; the two functions
// under test touch only Api, so the rest are present but inert.
const captured = [];
const ctx = {
  console,
  Date,
  Uint8Array,
  String,
  btoa: (s) => Buffer.from(s, 'binary').toString('base64'),
  TextDecoder,
  Map,
  Error,
  MvrZip: {},
  GdtfParse: {},
  Api: {
    importLibrary: async (doc, mode) => { captured.push({ doc, mode }); return { ok: true }; },
  },
};
ctx.globalThis = ctx;
vm.createContext(ctx);
vm.runInContext(fs.readFileSync(path.join(jsDir, 'mvrimport.js'), 'utf8'), ctx);
// `const MvrImport = ...` at file scope is NOT a context property, so reach
// it by evaluating its name — the harness gotcha this project already hit
// once in nodes_universe_test.js.
const MvrImport = vm.runInContext('MvrImport', ctx);

check(typeof MvrImport.libraryDocFromGdtf === 'function',
  'mvrimport.js does not export libraryDocFromGdtf — the shared builder is the whole point; ' +
  'without it each caller hand-rolls the record and they drift');
check(typeof MvrImport.rememberGdtfInLibrary === 'function',
  'mvrimport.js does not export rememberGdtfInLibrary');

// A fixture with three personalities, the ordinary case: an operator patches
// one of them and the other two are exactly what the library exists to keep.
const parsed = {
  manufacturer: 'GLP',
  model: 'JDC1',
  fixtureType: 'JDC1',
  modes: [
    { name: '24ch Extended', footprint: 24, channelFunctions: { 0: { attribute: 'Dimmer' } } },
    { name: '16ch Standard', footprint: 16, channelFunctions: {} },
    { name: '8ch Basic', footprint: 8, channelFunctions: {} },
  ],
  warnings: [],
};

// --- 1. the document carries every mode ------------------------------------
const doc = MvrImport.libraryDocFromGdtf(parsed, 'GLP@JDC1.gdtf', null);

check(doc.format === 'benny512-fixture-library',
  `doc.format = ${doc.format}, want benny512-fixture-library (the server rejects anything else)`);
check(doc.records && doc.records.length === 1, 'expected exactly one record');

const rec = (doc.records || [])[0] || {};
check(rec.manufacturer === 'GLP' && rec.model === 'JDC1',
  `record identity = ${rec.manufacturer}/${rec.model}, want GLP/JDC1`);

const names = (rec.modes || []).map(m => m.name);
check(rec.modes && rec.modes.length === 3,
  `record carries ${names.length} mode(s) [${names.join(', ')}], want all 3. ` +
  'A patch entry stores one mode, so any mode dropped here is unrecoverable later.');
['24ch Extended', '16ch Standard', '8ch Basic'].forEach(n => {
  check(names.includes(n), `mode "${n}" is missing from the library record`);
});
check((rec.modes || []).every(m => m.origin && m.origin.source === 'gdtf'),
  'every mode must carry gdtf provenance, or the library cannot tell a parsed ' +
  'manufacturer file from a profile harvested off a patch');
check((rec.modes || []).some(m => m.footprint === 24) && (rec.modes || []).some(m => m.footprint === 8),
  'footprints were not carried through with their modes');

// Channel functions ride along, not just names and counts — a mode without
// them cannot drive Rig Check's function-aware tests.
const ext = (rec.modes || []).find(m => m.name === '24ch Extended') || {};
check(ext.channelFunctions && ext.channelFunctions[0],
  'channelFunctions were dropped; a remembered mode that cannot drive Function check ' +
  'is a footprint, not a profile');

// --- 2. the original archive is retained when the bytes are supplied --------
check(!rec.sourceFiles,
  'sourceFiles present although no buffer was supplied — never invent an archive');

const withBytes = MvrImport.libraryDocFromGdtf(parsed, 'GLP@JDC1.gdtf', new Uint8Array([1, 2, 3, 4]).buffer);
const wrec = withBytes.records[0];
check(wrec.sourceFiles && wrec.sourceFiles.length === 1 && wrec.sourceFiles[0].name === 'GLP@JDC1.gdtf',
  'the original .gdtf archive was not retained when its bytes were available');
check(typeof (wrec.sourceFiles || [{}])[0].data === 'string',
  'retained archive is not base64 text');

// --- 3. remembering posts a merge, and never throws -------------------------
(async () => {
  captured.length = 0;
  const err = await MvrImport.rememberGdtfInLibrary(parsed, 'GLP@JDC1.gdtf', null);
  check(err === null, `rememberGdtfInLibrary returned an error on a good import: ${err}`);
  check(captured.length === 1, `expected one importLibrary call, got ${captured.length}`);
  check(captured[0] && captured[0].mode === 'merge',
    `import mode = ${captured[0] && captured[0].mode}, want "merge" — a GDTF import must never ` +
    'REPLACE the library, which would discard every other fixture type the user has collected');
  check((captured[0].doc.records[0].modes || []).length === 3,
    'the posted document lost modes between builder and call');

  // A library failure must be REPORTED, not thrown: it runs alongside an
  // import the operator actually asked for (patching entries), and losing
  // their patch to a library error would be the worse bug.
  ctx.Api.importLibrary = async () => { throw new Error('disk full'); };
  let threw = false;
  let msg = null;
  try {
    msg = await MvrImport.rememberGdtfInLibrary(parsed, 'x.gdtf', null);
  } catch (e) {
    threw = true;
  }
  check(!threw, 'rememberGdtfInLibrary threw on a library failure — that would abort the ' +
    'patch import the operator actually asked for');
  check(typeof msg === 'string' && msg.indexOf('disk full') >= 0,
    `expected the failure reported back for the status line, got ${JSON.stringify(msg)}`);

  // Nothing to remember is not an error, and must not post an empty record.
  captured.length = 0;
  ctx.Api.importLibrary = async (doc, mode) => { captured.push({ doc, mode }); return {}; };
  const none = await MvrImport.rememberGdtfInLibrary({ modes: [] }, 'empty.gdtf', null);
  check(none === null && captured.length === 0,
    'a GDTF with no modes posted a record anyway; an empty fixture type is noise in the library');

  if (failures) {
    console.error(`\n${failures} check(s) failed`);
    process.exit(1);
  }
  console.log('library_allmodes_test.js: all checks passed');
})();

File: internal/web/library_allmodes_test.go
package web

import (
	"bytes"
	"os"
	"os/exec"
	"regexp"
	"strings"
	"testing"
)

// TestLibraryRemembersAllGdtfModes runs library_allmodes_test.js, which
// exercises the literal mvrimport.js the browser loads.
//
// What it guards: a GDTF describes a fixture's whole personality list, but
// patching applies exactly one mode. `patch.Entry` stores a single mode
// (Mode, Footprint, ChannelFunctions), so the modes not patched exist only
// during the import — harvesting the show afterwards can never recover them.
// Every GDTF entry point therefore writes the FULL list to the library.
//
// Same Node prerequisite and skip-not-fail policy as the other JS suites:
// the code under test is the browser's own file, not a Go port of it.
func TestLibraryRemembersAllGdtfModes(t *testing.T) {
	nodePath, err := exec.LookPath("node")
	if err != nil {
		t.Skip("node not found on PATH — skipping library all-modes test")
	}

	cmd := exec.Command(nodePath, "static/js/testdata/library_allmodes_test.js")
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		t.Fatalf("library_allmodes_test.js failed: %v\n--- stdout ---\n%s\n--- stderr ---\n%s",
			err, stdout.String(), stderr.String())
	}
	t.Log(stdout.String())
}

// TestGdtfImportPathsAllReachTheLibrary is the coverage check the JS test
// cannot make: that every place a GDTF ENTERS the app actually calls the
// shared builder.
//
// The JS test proves the builder is correct. It cannot prove a fourth import
// path won't be added next month that quietly keeps only the applied mode —
// and that failure is invisible, because the patch looks right and the loss
// only shows up weeks later when the library turns out not to have the mode
// somebody needed.
//
// So this asserts against the literal screens: the three known entry points
// each reference the shared helpers, and none of them still hand-rolls a
// `format: 'benny512-fixture-library'` document of its own.
func TestGdtfImportPathsAllReachTheLibrary(t *testing.T) {
	read := func(name string) string {
		b, err := os.ReadFile("static/js/" + name)
		if err != nil {
			t.Fatalf("reading the literal %s the browser loads: %v", name, err)
		}
		return string(b)
	}

	mvrimport := read("mvrimport.js")
	library := read("library.js")
	patch := read("patch.js")

	// The builder and the best-effort wrapper both live in one place.
	for _, want := range []string{"function libraryDocFromGdtf", "async function rememberGdtfInLibrary"} {
		if !strings.Contains(mvrimport, want) {
			t.Errorf("mvrimport.js is missing %q — the shared GDTF→library path", want)
		}
	}

	// Every screen that takes in a GDTF uses it.
	for _, c := range []struct{ name, src, want, why string }{
		{"library.js", library, "MvrImport.libraryDocFromGdtf",
			"the Fixture Library dialog must build its record with the shared builder, " +
				"not the inline copy it used to carry"},
		{"patch.js (single GDTF)", patch, "MvrImport.rememberGdtfInLibrary",
			"applying a GDTF to patch entries must also remember the fixture type's other modes"},
	} {
		if !strings.Contains(c.src, c.want) {
			t.Errorf("%s does not call %s — %s", c.name, c.want, c.why)
		}
	}

	// The MVR path carries whole GDTF files; those modes must be surfaced
	// and remembered too, not left inside the parse cache.
	if !strings.Contains(mvrimport, "gdtfs") {
		t.Error("parseMvrFile does not surface the GDTFs it resolved, so an MVR import " +
			"cannot put their other modes in the library — the modes are parsed and thrown away")
	}
	if strings.Count(patch, "MvrImport.rememberGdtfInLibrary") < 2 {
		t.Error("patch.js remembers a GDTF in only one of its two import paths; an MVR " +
			"import carries GDTF files too, and their unpatched modes are lost at that moment")
	}

	// And nobody CONSTRUCTS the document shape any more. This is the actual
	// regression: two copies of one shape drifting apart.
	//
	// Matching on construction (`format: 'benny512-fixture-library'`) rather
	// than on any mention of the string, because library.js legitimately
	// VALIDATES it — `pending.doc.format !== 'benny512-fixture-library'` is
	// how it refuses a patch JSON handed to the library importer, and
	// deleting that check to satisfy a test would be a real regression.
	construct := regexp.MustCompile(`format\s*:\s*['"]benny512-fixture-library`)
	for _, c := range []struct{ name, src string }{
		{"library.js", library},
		{"patch.js", patch},
	} {
		if construct.MatchString(c.src) {
			t.Errorf("%s builds a library document inline. Use MvrImport.libraryDocFromGdtf — "+
				"a second hand-rolled copy is how the two drift, which is this project's "+
				"most-repeated defect", c.name)
		}
	}
	// The builder itself is the one place that may construct it.
	if !construct.MatchString(mvrimport) {
		t.Error("mvrimport.js's libraryDocFromGdtf no longer emits the library format marker")
	}
}
